### PR TITLE
Opendomain

### PR DESCRIPTION
**Pull request by JasonTheMain**
## Changes proposed in this pull request:
- Adding in module for opensearch, this has test values.
- added in iam policies that work with new elasticsearch and opensearch
-

## security considerations
More IAM policies, these only give access to elastic and open search options.

### CODE REVIEW

File: terraform/modules/opensearch_domain/opensearch.tf [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  engine_version = var.engine
+
+  cluster_config {
+    instance_type          = var.instance_type
+    zone_awareness_enabled = true
+  }
+
+  vpc_options {
+    subnet_ids = var.private_elb_subnets
+    #security_group_ids = var.bosh_security_group
```
Comment (markdboyd) [inline terraform/modules/opensearch_domain/opensearch.tf:12]: Can we get rid of the commented values?

File: terraform/modules/iam_role_policy/aws_broker/policy.json [hunk 2]
```diff
@@ -122,6 +122,48 @@
         "*"
       ]
     },
+    {
```
Comment (markdboyd) [inline terraform/modules/iam_role_policy/aws_broker/policy.json:125]: Why was this added?

File: terraform/modules/opensearch_domain/variables.tf [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  
+}
+variable "domain" {
+  
+}
+variable "private_elb_subnets" {
+  type = list(string)
+}
+
+#variable "bosh_security_group" {
```
Comment (markdboyd) [inline terraform/modules/opensearch_domain/variables.tf:14]: Can we remove this?

File: terraform/modules/opensearch_domain/variables.tf [hunk 4]
```diff
@@ -0,0 +1,15 @@
+variable "instance_type" {
+  
+}
+variable "engine" {
+  
+}
```
Comment (markdboyd) [inline terraform/modules/opensearch_domain/variables.tf:6]:
```suggestion
variable "engine" {
  type = string
  default = "Opensearch_2.5"
}
```

File: terraform/modules/iam_role_policy/aws_broker/policy.json (hunk 2 above)
Comment (JasonTheMain) [reply terraform/modules/iam_role_policy/aws_broker/policy.json]: I guess we could wait on this but our iam policies that state elasticsearch are going to deprecate. these commands should be the same but with the capability for opensearch.

File: terraform/stacks/main/stack.tf [hunk 5]
```diff
@@ -305,6 +305,15 @@ module "diego" {
   log_bucket_name = module.log_bucket.elb_bucket_name
 }
 
+module "opensearch" {
+  source = "../../modules/opensearch_domain"
+  instance_type= "t3.medium.search"
+  private_elb_subnets = [module.cf.services_subnet_az1, module.cf.services_subnet_az2]
+  #bosh_security_group = module.stack.bosh_security_group
+  engine= "OpenSearch_1.0"
+  domain= "test"
```
Comment (markdboyd) [inline terraform/stacks/main/stack.tf:314]:
```suggestion
  domain= "{var.stack_description}-logs-customer"
```

File: terraform/stacks/main/stack.tf [hunk 6]
```diff
@@ -305,6 +305,15 @@ module "diego" {
   log_bucket_name = module.log_bucket.elb_bucket_name
 }
 
+module "opensearch" {
+  source = "../../modules/opensearch_domain"
+  instance_type= "t3.medium.search"
+  private_elb_subnets = [module.cf.services_subnet_az1, module.cf.services_subnet_az2]
+  #bosh_security_group = module.stack.bosh_security_group
```
Comment (markdboyd) [inline terraform/stacks/main/stack.tf:312]: Please remove this

File: terraform/stacks/main/stack.tf [hunk 7]
```diff
@@ -305,6 +305,15 @@ module "diego" {
   log_bucket_name = module.log_bucket.elb_bucket_name
 }
 
+module "opensearch" {
```
Comment (markdboyd) [inline terraform/stacks/main/stack.tf:308]:
```suggestion
module "opensearch_logs_customer" {
```

We also need to make this module conditional on a boolean because we only want to test it in dev for now and not deploy in stage/prod. Here's an example of how to make a resource conditional:

https://github.com/cloud-gov/cg-provision/blob/df069274dbd9b6b0e5882c9ecb0c56143c82ac41/terraform/modules/iam_role/role.tf#L13

File: .github/dependabot.yml [hunk 8]
```diff
@@ -184,6 +184,10 @@ updates:
     directory: "/terraform/modules/logsearch"
     schedule:
       interval: "weekly"
+  - package-ecosystem: "terraform"
+    directory: "/terraform/modules/opensearcg_domain"
```
Comment (markdboyd) [inline .github/dependabot.yml:188]:
```suggestion
    directory: "/terraform/modules/opensearch_domain"
```

File: terraform/modules/opensearch_domain/opensearch.tf [hunk 9]
```diff
@@ -0,0 +1,13 @@
+resource "aws_opensearch_domain" "opensearch" {
```
Comment (markdboyd) [inline terraform/modules/opensearch_domain/opensearch.tf:1]: We need to add support for a few more things here:

- advanced security options: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/opensearch_domain#advanced_security_options. we should allow a variable for `internal_user_database_enabled` but default it to true. we should allow specifying a `master_user_name` and `master_user_password` via variables.

We should also add:

```
  node_to_node_encryption {
    enabled = true
  }
  
  encrypt_at_rest {
    enabled = true
  }  
```

File: terraform/stacks/main/stack.tf [hunk 10]
```diff
@@ -305,6 +305,14 @@ module "diego" {
   log_bucket_name = module.log_bucket.elb_bucket_name
 }
 
+module "opensearch_logs_customer" {
+  source = "../../modules/opensearch_domain"
+  instance_type= "t3.medium.search"
+  private_elb_subnets = [module.cf.services_subnet_az1, module.cf.services_subnet_az2]
+  engine= "OpenSearch_1.0"
```
Comment (markdboyd) [inline terraform/stacks/main/stack.tf:312]:
```suggestion
```

File: terraform/modules/opensearch_domain/variables.tf [hunk 11]
```diff
@@ -0,0 +1,13 @@
+variable "instance_type" {
+  
+}
+variable "engine" {
+  type = string
+  default = "Opensearch_2.5"
+}
+variable "domain" {
```
Comment (markdboyd) [inline terraform/modules/opensearch_domain/variables.tf:8]:
```suggestion
variable "domain_name" {
```

File: terraform/modules/opensearch_domain/opensearch.tf (hunk 9 above)
Comment (JasonTheMain) [reply terraform/modules/opensearch_domain/opensearch.tf:1]: sure!!

File: terraform/stacks/main/stack.tf [hunk 12]
```diff
@@ -305,6 +305,15 @@ module "diego" {
   log_bucket_name = module.log_bucket.elb_bucket_name
 }
 
+module "opensearch_logs_customer" {
+  source = "../../modules/opensearch_domain"
+  instance_type= "t3.medium.search"
+  private_elb_subnets = [module.cf.services_subnet_az1, module.cf.services_subnet_az2]
+  domain= "{var.stack_description}-logs-customer"
+  master_user_name = "test_user"
+  master_password = "CHANGEME"
```
Comment (markdboyd) [inline terraform/stacks/main/stack.tf:314]: Let's wire these up to come from the pipeline and make sure not to hardcode any values, even for testing

**Review by JasonTheMain — COMMENTED** [review]
Should be caught up with the comments. Thank you

File: terraform/stacks/main/stack.tf [hunk 13]
```diff
@@ -305,6 +305,15 @@ module "diego" {
   log_bucket_name = module.log_bucket.elb_bucket_name
 }
 
+module "opensearch_logs_customer" {
```
Comment (markdboyd) [inline terraform/stacks/main/stack.tf:308]: We need to make this module have a conditional `count` based on a boolean variable named `deploy_opensearch_logs_customer` or something

File: terraform/modules/opensearch_domain/opensearch.tf [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+  domain_name    = var.domain_name
+  engine_version = var.engine
+
+  advanced_security_options {
+    internal_user_database_enabled = true
```
Comment (markdboyd) [inline terraform/modules/opensearch_domain/opensearch.tf:6]: Let's make a variable for `internal_user_database_enabled` that defaults to true

File: terraform/modules/opensearch_domain/opensearch.tf [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+    master_user_options {
+      master_user_name     = var.master_user_name
+      master_user_password = var.master_user_password
+    }
+  }
+
+  cluster_config {
```
Comment (markdboyd) [inline terraform/modules/opensearch_domain/opensearch.tf:13]: We need to support these variables in this block:

- [dedicated_master_count](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/opensearch_domain#dedicated_master_count)
- [dedicated_master_enabled](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/opensearch_domain#dedicated_master_enabled)
- [dedicated_master_type](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/opensearch_domain#dedicated_master_type)
- [instance_count](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/opensearch_domain#instance_count)

File: terraform/stacks/main/stack.tf [hunk 16]
```diff
@@ -305,6 +305,15 @@ module "diego" {
...
+  source = "../../modules/opensearch_domain"
+  instance_type= "t3.medium.search"
```
Comment (markdboyd) [inline terraform/stacks/main/stack.tf:310]:
```suggestion
```

File: terraform/stacks/main/variables.tf [hunk 17]
```diff
@@ -74,6 +74,9 @@ variable "bucket_prefix" {
 
 variable "blobstore_bucket_name" {
 }
+variable "deploy_opensearch_logs_customer" {
+  default = "false"
```
Comment (markdboyd) [inline terraform/stacks/main/variables.tf:78]:
```suggestion
  default = false
```

File: terraform/stacks/main/stack.tf [hunk 18]
```diff
@@ -305,6 +305,15 @@ module "diego" {
   log_bucket_name = module.log_bucket.elb_bucket_name
 }
 
+module "opensearch_logs_customer" {
+  count                = var.deploy_opensearch_logs_customer ? 1 : 0
+  source               = "../../modules/opensearch_domain"
+  private_elb_subnets  = [module.cf.services_subnet_az1, module.cf.services_subnet_az2]
+  domain_name          = "{var.stack_description}-logs-customer"
+  master_user_name     = "{var.master_user_name}"
+  master_user_password = "{var.master_user_password}"
```
Comment (markdboyd) [inline terraform/stacks/main/stack.tf:314]:
```suggestion
  master_user_name     = var.opensearch_logs_customer_master_username
  master_user_password = var.opensearch_logs_customer_master_password
```

While this is passing terraform validation, it won't do what you expect on `terraform plan`. You're interpolating a variable of `master_user_name` here, but there is no variable defined in `terraform/stacks/main/variables.tf` with that name. 

So we should add these variables to that file:

- `opensearch_logs_customer_master_username`
- `opensearch_logs_customer_master_password`

File: terraform/modules/opensearch_domain/variables.tf [hunk 19]
```diff
@@ -0,0 +1,37 @@
+variable "instance_type" {
```
Comment (markdboyd) [inline terraform/modules/opensearch_domain/variables.tf:1]: We should define a `type` on all of our variables so Terraform can validate the values that we pass in

File: ci/pipeline.yml [hunk 20]
```diff
@@ -388,11 +388,13 @@ jobs:
       TF_VAR_smtp_ingress_cidr_blocks: '["((staging_private_cidr_1))", "((staging_private_cidr_2))", "((production_private_cidr_1))", "((production_private_cidr_2))"]'
       TF_VAR_restricted_ingress_web_cidrs: ((tooling_restricted_ingress_web_cidrs))
       TF_VAR_restricted_ingress_web_ipv6_cidrs: ((tooling_restricted_ingress_web_ipv6_cidrs))
-      TF_VAR_blobstore_bucket_name: bosh-tooling-blobstore
+      TF_VAR_blobstore_bucket_name: bosh-tooling-blobsxtore
```
Comment (markdboyd) [inline ci/pipeline.yml:391]:
```suggestion
      TF_VAR_blobstore_bucket_name: bosh-tooling-blobstore
```

File: terraform/stacks/main/stack.tf [hunk 21]
```diff
@@ -305,6 +305,15 @@ module "diego" {
   log_bucket_name = module.log_bucket.elb_bucket_name
 }
 
+module "opensearch_logs_customer" {
+  count                = var.deploy_opensearch_logs_customer ? 1 : 0
+  source               = "../../modules/opensearch_domain"
+  private_elb_subnets  = [module.cf.services_subnet_az1, module.cf.services_subnet_az2]
+  domain_name          = "{var.stack_description}-oc" #OC = opseaarch-customer
```
Comment (markdboyd) [inline terraform/stacks/main/stack.tf:312]:
```suggestion
  domain_name          = "{var.stack_description}-logs-customer" 
```

**Review by markdboyd — APPROVED** [review]
